File: September2022/C++/BagOfTokens.cpp
// https://leetcode.com/problems/bag-of-tokens/
class Solution {
public:
    // TC: O(nlogn)
    // SC: O(1)
    int bagOfTokensScore(vector<int>& tokens, int power) {
        int score = 0;
        int maxScore = 0;
        
        // sort tokens in ascending order
        sort(tokens.begin(), tokens.end());
        
        int left = 0;
        int right = tokens.size() - 1;
        
        while (left <= right) {
            if (power >= tokens[left]) {                // option 1
                power -= tokens[left];
                score++;
                left++;
                maxScore = max(maxScore, score);
            } else if (left != right && score >= 1) {   // option 2
                power += tokens[right];
                score--;
                right--;
            } else {
                break;
            }
        }
        
        return maxScore;
    }
};

File: September2022/Java/BagOfTokens.java

// @saorav21994
// TC : O(nlogn)
// SC : O(1)

class Solution {
    public int bagOfTokensScore(int[] tokens, int P) {
        
        Arrays.sort(tokens);
        int s = 0;
        int n = tokens.length;
        if (n == 0) {
            return s;
        }
        else if (tokens[0] > P) {
            return s;
        }
        else {
            int i = 0, j = n-1;
            while (i <= j) {
                if (tokens[i] <= P) {
                    P -= tokens[i];
                    s += 1;
                    i += 1;
                }
                else {
                    if (i < j && s > 0) {
                        P += tokens[j];
                        j -= 1;
                        s -= 1;
                    }
                    else {
                        break;
                    }
                }
            }
        }
        return s;
        
    }
}
